File: libs/power-log-parser/src/lib/cards/mimicry.ts
import { GameTag } from '@firestone-hs/reference-data';
import { Action } from '../models/action';
import { FullEntity } from '../models/entity';
import { Node, NodeType } from '../models/node';
import type { GameState } from '../state/game-state';
import type { StateFacade } from '../state/state-facade';

export class Mimicry {
	static PredictCardId(
		gameState: GameState,
		entityId: number,
		creatorEntityId: number,
		node: Node,
		stateFacade: StateFacade | null,
	): string | null {
		if (node.Parent?.Type !== NodeType.Action) {
			return null;
		}

		const act = node.Parent.Object as Action;
		const candidates = act.Data
			.filter((d): d is FullEntity => d instanceof FullEntity)
			.map((d) => gameState.CurrentEntities.get(d.Entity))
			.filter((d) => d != null && d.GetTag(GameTag.COPIED_FROM_ENTITY_ID) === entityId);

		const last = candidates.length > 0 ? candidates[candidates.length - 1] : null;
		return last?.CardId ?? null;
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Events/Cards/Mimicry.cs
using HearthstoneReplays.Enums;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using HearthstoneReplays.Parser;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Action = HearthstoneReplays.Parser.ReplayData.GameActions.Action;
using HearthstoneReplays.Parser.ReplayData.Entities;

namespace HearthstoneReplays.Events.Cards
{
    internal class Mimicry
    {
        public static string PredctCardId(GameState gameState, int entityId, int creatorEntityId, Node node, StateFacade stateFacade)
        {
            if (node.Parent?.Type != typeof(Action))
            {
                return null;
            }

            var act = node.Parent.Object as Action;
            return act.Data
                .Where(d => d is FullEntity)
                .Select(d => d as FullEntity)
                .Select(d => gameState.CurrentEntities.GetValueOrDefault(d.Entity))
                .Where(d => d.GetTag(GameTag.COPIED_FROM_ENTITY_ID) == entityId)
                .LastOrDefault()?.CardId;
        }
    }
}
